docs/source/working/data.rst

```diff
@@ -0,0 +1,148 @@
+These basic data types serve most needs for the majority of applications, but more specific solutions may be useful or even necessary.
+In the next sections, we will explain :ref:`how a new data type can be created<working_data_creating_new_types>` and what :ref:`guidelines<working_data_design_guidelines>` should ideally be observed during the design process.
+
+
+.. _working_data_creating_new_types:
+
+Creating new data types
+=======================
+
+Creating a new data type is as simple as creating a new sub class of the base :class:`~aiida.orm.nodes.data.data.Data` class.
+
+.. code:: python
+
+    from aiida.orm import Data
+
+    class NewData(Data)
+        """A new data type that wraps a single value."""
+
+At this point, our new data type does nothing special.
+Typically, one creates a new data type to represent a specific type of data.
+For the purposes of this example, let's assume that the goal of our ``NewData`` type is to store a single numerical value.
+To allow one to construct a new ``NewData`` data node with the desired ``value``, for example:
+
+.. code:: python
+
+    node = NewData(value=5)
+
+we need to allow passing that value to the constructor of the node class.
+Therefore, we have to override the constructor :meth:`~aiida.orm.nodes.node.Node.__init__`:
+
+.. code:: python
+
+    from aiida.orm import Data
+
+    class NewData(Data)
+        """A new data type that wraps a single value."""
+
+        attribute_name = 'value'
+
+        def __init__(self, **kwargs)
+            value = kwargs.pop(self.attribute_name)
+            super(NewData, self).__init__(**kwargs)
+
+.. warning::
+
+    For the class to function properly, the signature of the constructor **cannot be changed** and the constructor of the parent class **has to be called**.
+
+Before calling the construtor of the base class, we have to remove the ``value`` keyword from the keyword arguments ``kwargs``, because the base class will not expect it and will raise an exception if left in the keyword arguments.
+The final step is to actually *store* the value that is passed by the caller of the constructor.
+The previous snippet removes it from the ``kwargs`` but then does not do anything with it.
+A new node has two locations to permanently store any of its properties:
+
+* the database
+* the file repository
+
+The section on :ref:`design guidelines<working_data_design_guidelines>` will go into more detail what the advantages and disadvantages of each option are and when to use which.
+For now, since we are storing only a single value, the easiest and best option is to use the database.
+Each node has *attributes* that can store any key-value pair, as long as the value is JSON serializable.
+We can use these attributes to store the ``value`` of our ``NewData`` data type:
+
+.. code:: python
+
+    from aiida.orm import Data
+
+    class NewData(Data)
+        """A new data type that wraps a single value."""
+
+        attribute_name = 'value'
```

Comment:
I wasn't sure about this one either. I was worried that using `value` for the attribute and `value` for the attribute name, the reader might mistakenly think they have to be the same name. I wanted to convey that it is just a name that could be anything they want. Since the user of the class should never need to know how data is stored by the class internally. But with this simple example that is difficult to get across. Anyway, I have removed it for now, I think you are right that it will be better.